src/halmos/__main__.py

```diff
@@ -504,7 +523,123 @@ def run(
     return 0 if passed else 1
 
 
-def gen_model_from_sexpr(args: argparse.Namespace, idx: int, sexpr: str) -> ModelWithContext:
+@dataclass(frozen=True)
+class SetupAndRunSingleArgs:
+    hexcode: str
+    abi: List
+    setup_info: FunctionInfo
+    fun_info: FunctionInfo
+    args: argparse.Namespace
+
+
+def setup_and_run_single(fn_args: SetupAndRunSingleArgs) -> int:
+    try:
+        setup_ex = setup(
+            fn_args.hexcode,
+            fn_args.abi,
+            fn_args.setup_info,
+            fn_args.args,
+        )
+    except Exception as err:
+        print(color_warn(f'Error: {fn_args.setup_info.sig} failed: {type(err).__name__}: {err}'))
+        if args.debug: traceback.print_exc()
+        return SETUP_FAILED
+
+    try:
+        exitcode = run(
+            setup_ex,
+            fn_args.abi,
+            fn_args.fun_info,
+            fn_args.args,
+        )
+    except Exception as err:
+        print(f'{color_warn("[SKIP]")} {fn_args.fun_info.sig}')
+        print(color_warn(f'{type(err).__name__}: {err}'))
+        if args.debug: traceback.print_exc()
+        return TEST_FAILED
+
+    return exitcode
+
+
+def extract_setup(methodIdentifiers: Dict[str, str]) -> FunctionInfo:
+    setup_sigs = sorted([ (k,v) for k,v in methodIdentifiers.items() if k == 'setUp()' or k.startswith('setUpSymbolic(') ])
+
+    if not setup_sigs:
+        return FunctionInfo()
+
+    (setup_sig, setup_selector) = setup_sigs[-1]
+    setup_name = setup_sig.split('(')[0]
+    return FunctionInfo(setup_name, setup_sig, setup_selector)
+
+
+@dataclass(frozen=True)
+class RunArgs:
+    # signatures of test functions to run
+    funsigs: List[str]
+
+    # code of the current contract
+    hexcode: str
+
+    abi: List
+    methodIdentifiers: Dict[str, str]
+
+
+def run_parallel(run_args: RunArgs) -> Tuple[int, int]:
+    hexcode, abi, methodIdentifiers = run_args.hexcode, run_args.abi, run_args.methodIdentifiers
+
+    setup_info = extract_setup(methodIdentifiers)
+    if args.verbose >= 1:
+        print(f'Running {setup_info.sig}')
+
+    fun_infos = [FunctionInfo(funsig.split('(')[0], funsig, methodIdentifiers[funsig]) for funsig in run_args.funsigs]
+    single_run_args = [SetupAndRunSingleArgs(hexcode, abi, setup_info, fun_info, args) for fun_info in fun_infos]
+
+    # dispatch to the shared process pool
+    exitcodes = process_pool.map(setup_and_run_single, single_run_args)
+
+    num_passed = sum(1 for x in exitcodes if x == 0)
+    num_failed = sum(1 for x in exitcodes if x != 0)
```

Comment:
at this point, `num_failed` is always 0, because `existcodes` becomes empty after the lazy iterables have already been consumed when computing `num_passed`.

you can reproduce this by running two tests in parallel where one of them fails.

Reply:
oh damn, great find 😱

fixed in 910a1cb